Integrate `websockets` protocol support (#119)
ballerina-platform/asyncapi-tools: docs/spec/spec.md

# Specification: Ballerina AsyncAPI for Websockets

> _Authors_: @HasithaAthukorala @thushalya  
> _Reviewers_: @bhashinee @shafreenAnfar
> _Created_: 2024/06/27   
> _Updated_: 2024/06/28
> _Edition_: Swan Lake

Reviewer: ```suggestion
_Authors_: @HasithaAthukorala @thushalya  
_Reviewers_: @bhashinee @shafreenAnfar   
_Created_: 2024/06/27   
_Updated_: 2024/06/28   
_Edition_: Swan Lake   
```

- If there are multiple services in the project, the tool will generate a separate specification for each service, by equating each service to a channel.
- The generated specification is of version 2.5.0.

### How the AsyncAPI specification is generated

#### Info Object

```yaml
info:
  title: <service_name> 
  version: <package_version>
```

The `service_name` is derived from the name of the service. If the name of the service is `/`, the `service_name` will be derived from the file name. Otherwise, the `service_name` will be the name of the service.

The `package_version` is derived from the version defined in the `Ballerina.toml` file, if available. If the version is not defined in the `Ballerina.toml` file, the `package_version` will be `0.0.0`.

#### Server Object

```yaml
servers:
  "development":
    url: "{server}:{port}/"
    protocol: ws
    protocolVersion: "13"
    variables:
      server:
        default: ws://localhost
      port:
        default: "9092"
```

The server name will be `development` by default, which can be changed by the user once the specification is generated. 

> The `server` URL will be `ws://localhost`, and if the `websocket:Listener` has SSL configurations, the `server` URL will be `wss://localhost`. The `port` will be derived from the port number that the `websocket:Listener` is attached to. The protocol version will be `13` by default. The `protocol` will be `ws`.

Reviewer: ```suggestion
The `server` URL will be `ws://localhost`, and if the `websocket:Listener` has SSL configurations, the `server` URL will be `wss://localhost`. The `port` will be derived from the port number that the `websocket:Listener` is attached to. The protocol will be `ws`, and the WebSocket protocol version used will be `13`.
```

#### Channel Object

```yaml
channels:
  /:
    description: An echo service that echoes the messages sent by the client.
    subscribe:
      message:
        $ref: '#/components/messages/Response'
    publish:
      message:
        $ref: '#/components/messages/Hello'
```

Currently, this tool supports only one channel per specification. If there are multiple services in the project, the tool will generate a separate specification for each service, by equating each service to a channel.
The equivalent Ballerina service class is given below. The `description` will be the doc comment provided on the `resource` function of the service.

```Ballerina
@websocket:ServiceConfig{dispatcherKey: "event"}
service / on websocketListener {
    # An echo service that echoes the messages sent by the client.
    # + return - User status
    resource function get .() returns websocket:Service|websocket:UpgradeError {
        return new WsService();
    }

> 

Reviewer: ```suggestion
```

When there are query parameters in the `resource` function, the `bindings` object will be added to the `channel` object with the relevant query parameters under the `query` object.

When there are headers in the `resource` function, the `bindings` object will be added to the `channel` object with the relevant headers under the `headers` object.

Here, the `bindingsVersion` will be same as the `version` in the `info` object.

The returned `WsService` class is as follows.

```Ballerina
service class WsService {
    *websocket:Service;

    remote function onHello(Hello clientData) returns Response? {
        return {message:"You sent: " + clientData.message};
    }

}
```

The `subscribe` and `publish` objects will be derived from the `remote` functions available in the service class.
When there are multiple `subscribe` and `publish` payload types, the tool will add them all with the `oneOf` property.

#### Components Object

The `components` object will contain the `messages` and `schemas` objects. `messages` refer to what you can `subscribe` to and `publish` in the channel. `schemas` refer to the data types used in `messages`.

> `publish` messages must always be a `record` type of the name of the `remote` function without the `on` word. For example, if the `remote` function is `onHello`, the `subscribe` message type shall be `Hello`.

Reviewer: ```suggestion
`publish` messages must always be a `record` type of the name of the `remote` function without the `on` word. For example, if the `remote` function is `onHello`, the `publish` message type shall be `Hello`.
```

## 3. Generating the Ballerina client

This tool enables the generation of a websocket client based on a given AsyncAPI specification. It can be done by running the following command.

```bash
bal asyncapi --protocol ws -i <input_file> --o <output_directory>
```

| Command Parameter |                           Description                            | Mandatory |    Default Value    |
|:-----------------:|:----------------------------------------------------------------:|:---------:|:-------------------:|
|    --protocol     |                   indicates the protocol used.                   |    No     |        http         |
|   <-i, --input>   |         specifies the file containing the specification          |    Yes    |          -          |
|  <-o, --output>   |                  specifies the output directory                  |    No     | <current_directory> |
|   --with-tests    |          specifies whether to generate a test skeleton           |    No     |        false        |
|     --license     | specifies the path for a text file containing the license header |    No     |          -          |

> The command generates an AsyncAPI specification for the given service running on websockets. The generated client will be generated in the specified output directory. If the output directory is not specified, the client will be generated in the current directory.

Reviewer: ```suggestion
The command generates an AsyncAPI specification for the given service running on WebSockets. The generated client will be generated in the specified output directory. If the output directory is not specified, the client will be generated in the current directory.
```
